File: leetcode/0740-delete-and-earn/solution.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        if (nums.empty()) return 0;

        int maxVal = *max_element(nums.begin(), nums.end());
        vector<long long> points(maxVal + 1, 0);
        for (int x : nums) {
            points[x] += x;
        }

        long long take = 0, skip = 0;
        for (int v = 1; v <= maxVal; v++) {
            long long newTake = skip + points[v];
            long long newSkip = max(take, skip);
            take = newTake;
            skip = newSkip;
        }

        return (int)max(take, skip);
    }
};

File: leetcode/0740-delete-and-earn/Solution.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }

        int maxVal = 0;
        for (int x : nums) {
            maxVal = Math.max(maxVal, x);
        }

        long[] points = new long[maxVal + 1];
        for (int x : nums) {
            points[x] += x;
        }

        long take = 0, skip = 0;
        for (int v = 1; v <= maxVal; v++) {
            long newTake = skip + points[v];
            long newSkip = Math.max(take, skip);
            take = newTake;
            skip = newSkip;
        }

        return (int) Math.max(take, skip);
    }
}
